### PR TITLE
MAGECLOUD-3598: Set Varnish as default cache

### PR DESCRIPTION
**Pull request by NadiyaS**
### Description
Set up Varnish as default cache instead of MySQL for Cloud Docker
Issue link: [MAGECLOUD-3598](https://magento2.atlassian.net/browse/MAGECLOUD-3598)

### Manual testing scenarios
**Precondition**
Build PHP:
```bash
docker image build -t test/php-cli:7.3 vendor/magento/magento-cloud-docker/images/php/7.3-cli/
```
^ Replace `7.3` with appropriate version.

Add `docker-compose.override.yml`

```yaml
version: '2.1'
services:
    deploy:
        image: 'test/php-cli:7.3'
```
^ Replace `7.3` with appropriate version

**Scenarios**: 

**1.** Base scenario
- install Magento on the Docker environment 
- check that Magento uses Varnish for FPC:
    - FPC is Varnish in Magento Admin interface
    - `env.php` file has `http_cache_hosts` as `varnish`
    - open Magento Storefrons page. It should use Varnish.

**2.** Custom data not updated
- add mysql for FPC in config.php file using CLI command ` php bin/magento config:set system/full_page_cache/caching_application 1 --lock-env`
- remove `http_cache_hosts` from `env.php` file
- run deployment again
- check that Magento uses Mysql for FPC
- `env.php` file does not have `http_cache_hosts`

**3.** No config changes for Magento 2.1.x
- install Magento 2.1.x
- check that after deployment Varnish configuration was not set 

**4.** Varnish not installed
 - install magento on the Docker environment without varnish. For this add `--no-varnish` option.
`vendor/bin/ece-docker build:compose --no-varnish`
 - check that after this varnish image does not exist
 - check that after deployment Varnish configuration was not set and Magento uses MySQL for FPC

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages

releasenotes
Updated the Varnish service configuration:
- The Varnish service is now provisioned by default when you deploy a Magento Cloud Docker environment using a supported version of the Magento Cloud application template version 2.2.0 or later.
- Added the `--no-varnish` option to the `docker:build` command to skip Varnish service installation when you generate the configuration for a Magento Cloud Docker environment .

### CODE REVIEW

File: images/php/7.1-cli/bin/cloud-deploy [hunk 1]
```diff
@@ -8,4 +8,18 @@ echo "Running \"deploy\" hook."
 
 run-hooks deploy
 
+if [[ "$MAGENTO_RUN_MODE" == "production" ]]; then
```
Comment (shiftedreality) [inline images/php/7.1-cli/bin/cloud-deploy]: Provided this implementation still does not track if Varnish is present. Should we consider it as a required service on Cloud Docker?

File: images/php/7.1-cli/bin/cloud-deploy [hunk 2]
```diff
@@ -8,4 +8,18 @@ echo "Running \"deploy\" hook."
...
+    minor_magento_version=$(magento-command --version | sed "s/Magento CLI \([0-9]*\.[0-9]*\).*/\1/")
+    if [[ "$minor_magento_version" != "2.1" ]]; then
+        echo "Setting Varnish for FPC."
+        output=$(magento-command config:set system/full_page_cache/caching_application 2)
```
Comment (shiftedreality) [inline images/php/7.1-cli/bin/cloud-deploy:15]: Adding `--lock` would ensure the value won't get overwritten, and error code here is only expected if exception occured. Can be re-written in 

```
magento-command config:set system/full_page_cache/caching_application 2 --lock && magento-command setup:config:set --http-cache-hosts=varnish -n
```

**Review by shiftedreality — CHANGES_REQUESTED** [review]

File: images/php/7.1-cli/bin/cloud-deploy (hunk 2 above)
Comment (NadiyaS) [reply images/php/7.1-cli/bin/cloud-deploy:15]: Actually, this value was added to DB for providing possibility to change this value from Admin UI.
And if this value exists in files we won't change it.

File: images/php/7.1-cli/bin/cloud-deploy (hunk 1 above)
Comment (NadiyaS) [reply images/php/7.1-cli/bin/cloud-deploy]: We have this service by default, isn't it.

Comment (michaelcasey316) [reply images/php/7.1-cli/bin/cloud-deploy]: Another possibility for testing for the presence of Varnish:

1. Change the Varnish config to emit headers that reveal when Varnish is in use (currently it hides everything).
2. Use cURL to read the headers from port 80/443 and check for the presence of the Varnish header.
3. Change the config depending on the results of the request.

**Review by michaelcasey316 — COMMENTED** [review]
I approve of the changes, though I'm not sure this satisfies the concern that someone might accidentally enable Varnish when it's not running in Docker. Will Magento set `full_page_cache/caching_application` to  `2` even when the Varnish server is not actually running?

**Comment by shiftedreality** [comment]
Comment after latest public grooming:

We agreed to make varnish optional with `--no-varnish` option on both Developer and Production modes.
This will remove varnish as a service from resulting `docker-compose`.

The script should be able to detect the Varnish presence.

**Review by shiftedreality — APPROVED** [review]

**Review by shiftedreality — DISMISSED** [review]
Overall looks good, you'll have to resolve conflicts after https://github.com/magento/magento-cloud-docker/pull/113

**Review by shiftedreality — DISMISSED** [review]

**Review by oshmyheliuk — DISMISSED** [review]

**Review by shiftedreality — DISMISSED** [review]

**Review by oshmyheliuk — DISMISSED** [review]

**Comment by YPyltiai** [comment]
@aepod , @joeshelton-wagento , please take a look. This is the one we discussed during grooming.

**Comment by andriyShevtsov** [comment]
QA approved

There no way to test flow with magento 2.1.x because of this issue https://magento2.atlassian.net/browse/MAGECLOUD-4979 so decided to skip it.

**Review by shiftedreality — APPROVED** [review]

**Review by oshmyheliuk — APPROVED** [review]

**Pull request merged**